3단계 - 로또(자동)

popotec opened the pull request:
안녕하세요!
3단계 구현 제출 합니다.

프로그램을 설계하며 클래스를 분리하고 역할을 어떻게 나눌지 고민을 많이 한 결과물인데..
제대로 역할 분리를 잘 한 것인지 확신은 안듭니다.

고견 부탁드립니다..!

popotec committed:
refactor: 코드리뷰 반영

1. @NullSource, @EmptySource을 @NullAndEmptySource 하나로 변경
2. 불필요 주석 삭제
3. 불필요 exception signature 삭제
4. Pattern 객체 생성을 클래스 변수에서 선언
5. RunTimeException 예외를 구체화한 예외로 변경
6. sum 구할때 stream collect 사용
7. Parser 클래스를 정의하여 입력 문자열을 패턴을 적용해 파싱하는 것을 별도의 역할로 분리
8. vadlidateCustomDelimiter 메서드 로직 변경

popotec committed:
docs : 3단계 프로그램 정의

popotec committed:
refactor : 패키지 분리

popotec committed:
feat : 랜덤 로또 번호 생성 기능

popotec committed:
fix : 로또 번호 생성 방법 변경

- 랜덤 클래스를 이용해 생성하면 중복된 숫자의 경우를 체크하고 다시 뽑도록 해야하는 문제가 존재
=> List에 로또 번호가 될 수 있는 숫자들을 미리 저장해놓고 Collections.shuffle을 이용해 순서를 섞은후 앞에서부터 6개를 pick

popotec committed:
feat : 로또 티켓 발급

6자리 번호를 가진 로또 티켓 발급

popotec committed:
refactor : 정적 팩토리 메서드로 LotteryTicket을 생성하도록 변경

- (이전) Client가 LotteryNumberMaker에 대한 정보도 알고 LotteryTicket에 대한 정보도 알아야 LotteryNumberMaker로 6자리 숫자를 생성 후 LotteryTicket 생성
- (변경) Client가 LotteryTicket의 정적 팩토리 매서드만 호출하면 티켓 생성이 가능하므로, 클라이언트가 티켓의 내부 생성 로직을 몰라도 됨.

popotec committed:
feat : 구매 금액 유효성 체크

- 금액이 티켓 단위 가격 미만인 경우 예외처리
- 금액이 티켓 단위 가격으로 나누어떨어지지 않는 경우 예외처리

popotec committed:
feat : 구매 금액을 문자열로 입력받은 경우 처

- int형으로 변환하여 처리
- int형 변환시 문제가 있는 경우(숫자가 아닌 문자가 포함되어있는경우) 예외 처리

popotec committed:
feat : 구매 금액을 유효성 검증을 만족하지 않는 예외

popotec committed:
docs : 예외 클래스들에 설명 주석 추가

popotec committed:
refactor : 일급콜렉터, 원시값 포장 적용

- LotterNumber 클래스로 원시값 포장
- List 타입으로 저장한 6개의 로또 번호를 LotteryNumberGroup으로 변환

popotec committed:
refactor : 클래스명 일관성있게 변경

Lottery, Lotter -> Lotto로 변경

popotec committed:
refactor : 클래스명 일관성있게 변경

Lottery, Lotter -> Lotto로 변경

popotec committed:
chore: 로또번호 최소값 최대값 변수 접근제어자 변경(private->protected)

같은 패키지내에서 값을 공유하게 하여 LottoNumber에서 동일한 변수를 선언하지 않고 가져다 쓰도록 함

popotec committed:
feat: 원시타입을 포장한 LottoNumber의 생성 및 유효성 검증

popotec committed:
refactor: LottoNumberGroup으로 Wrapping 하지 않고 LottoTicket 자체가 일급콜렉터 역할을 하도록 수정

popotec committed:
refactor: LottoNumberGroup으로 Wrapping 하지 않고 LottoTicket 자체가 일급콜렉터 역할을 하도록 수정

popotec committed:
feat: 정수형 리스트로 당첨번호를 입력받기

popotec committed:
feat: 당첨번호 입력 데이터 예외 처리

1. 중복된 숫자가 존재하는 경우 예외
2. 갯수가 6개가 아닌 경우 예외(많은경우 or 적은경우)

popotec committed:
chore: 예외 클래스 상속관계 설정

popotec committed:
test: 구체적인 예외 클래스로 변경

popotec committed:
style: 불필요 개행 제거

popotec committed:
feat: 비교연산을 위한 eqauls, hashcode 구현

popotec committed:
refactor: 로또 갯수 상수화

popotec committed:
feat: 문자열을 입력받아 로또 당첨 번호 생성

popotec committed:
test : ArgumentsAccessor를 이용하여 배열을 파라미터로 테스트

popotec committed:
chore : 당첨번호 입력 문자열 형식이 올바르지 않을 경우 에러 메세지 변경

popotec committed:
refactor : List<LottoTicket>을 일급컬렉션 LottoTickets로 래핑

popotec committed:
refactor : LottoTicket의 유연한 생성을 위해 정적 팩토리 패턴 대신 복수의 생성자로 변경

popotec committed:
feat : 정답과 로또티켓을 비교하여 맞춘 번호 갯수 리턴

popotec committed:
feat : 정답과 로또티켓들 비교 결과 생성

popotec committed:
feat : 정답과 로또티켓들 비교 결과 생성

popotec committed:
chore : 메서드명 변경

popotec committed:
feat: 당첨 종류와 갯수에 따른 수익률 계산

popotec committed:
feat: 당첨 종류와 갯수에 따른 수익률 계산

popotec committed:
feat: toString 오버라이딩 구현

popotec committed:
feat: (UI연계) 구매금액을 입력받아 티켓 생성 후 화면에 출력

popotec committed:
feat: (UI연계) 구매금액을 입력받아 티켓 생성 후 화면에 출력

popotec committed:
refactor: 예외 클래스 상속관계 리팩토링

popotec committed:
test: 입력 금액이 유효한 최대 금액을 초과하는 경우 예외 처리

popotec committed:
refactor: 값 상수화

popotec committed:
refactor: 예외 성격 변경

popotec committed:
refactor: 불필요 예외 클래스 삭제

popotec committed:
feat: 사용자에게 보여주는 메세지 출력 담당

popotec committed:
feat: 로또 프로그램 실행과 제어

popotec committed:
docs : 기능정의 추가

popotec committed:
refactor : 패키지 구조 분리

popotec committed:
style : 예외 클래스 주석 추가

popotec committed:
refactor : lottoNumberCandidate를 Integer 리스트에서 LottoNumber 리스트로 변경

popotec committed:
docs : 클래스 설명 주석 추가

popotec committed:
refactor : 로또 번호 갯수를 LottoNumbersMaker에 public static 접근제어로 선언하여 사용.

클래스 내에서 로또 갯수를 별도 선언하면 동일 값을 여러 군데에서 선언하게 되어 만약에 로또 갯수가 수정될 경우 수정포인트가 여러개 발생하므로 변경

popotec committed:
docs : 클래스 설명 주석 추가

popotec committed:
docs : 기능 정의 보완

popotec committed:
chore : 입력금액 변수명 변경

money -> inputAmount

popotec committed:
fix : 로또 번호들 오름차순 정렬

popotec committed:
chore : 불필요 개행 삭제

popotec committed:
test : 로또 당첨 결과 RANK enum 테스트 구현

popotec committed:
test : 티켓 갯수 확인 테스트

popotec committed:
test : 10번의 반복 테스트에서 유효한 티켓이 생성되는지 테스트

popotec committed:
chore : 테스트 메서드명 변경

popotec committed:
chore : 접근제어자 변경

외부에서 해당 함수를 직접 접근할 필요 없으므로 private으로 변경

popotec committed:
refactor : 출력 메세지 상수화

popotec committed:
docs : 목록을 나타내는 문자 추가

minsungryu commented on src/main/java/lottoservice/LottoMain.java:
```diff
@@ -0,0 +1,76 @@
+
+	private String inputAmountForBuyLotto() {
+		ResultView.printGuideMessage(GUIDE_MESSAGE_ENTER_INPUT_AMOUNT);
+		return InputView.readLine();
```
InputView, ResultView의 구현이 콘솔에 의존적인데요,
만약 요구사항이 바뀌어서 이 로또 게임을 웹 서비스로 제공해야할 경우 수정할 부분이 많아질 것 같습니다.

이처럼 관심사 밖의 변동사항에 휘둘리지 않기 위해 DIP(의존성 역전 원칙)를 활용할 수 있는데요,
InputView, ResultView는 행위 중심의 인터페이스로 선언한다면 로또 게임에서는 단순히 외부로부터 입력을 받고 결과를 출력해주는 추상적인 행위에만 의존하게 되며 실제로 그 일은 구현체가 담당하게 됩니다.
이를테면 현재 코드를 ConsoleInputView, ConsoleResultView와 같은 구현체로 만들 수 있겠네요.

minsungryu commented on src/main/java/lottoservice/LottoMain.java:
```diff
@@ -0,0 +1,76 @@
+
+public class LottoMain {
+
+	private static String GUIDE_MESSAGE_ENTER_INPUT_AMOUNT = "구매금액을 입력해주세요.";
```
꼼꼼하게 상수를 선언해주셨네요 👍

minsungryu commented on src/main/java/lottoservice/exception/UnexpectedMatchException.java:
```diff
@@ -0,0 +1,12 @@
+ * 당첨번호와 티켓의 비교 과정에서 로직상 발생할 수 없는 Null이 발생한 경우 예외
+ */
+
+public class UnexpectedMatchException extends RuntimeException{
```
저번에 제가 구체적인 예외를 던지면 좋겠다고 코멘트를 남겼었는데
그래서인지 예외 클래스를 꼼꼼하게 잘 나눠주셨네요. 힘들지 않으셨나 모르겠네요 ㅎㅎ

현실적으로는 예외 클래스가 많이 생기면 관리하기 힘들기 때문에
이미 자바에서 제공하는 많은 예외 클래스 중 적합한 것을 선택하고 메시지를 잘 전달하는 것만으로도 충분할 수 있습니다.(IllegalArgumentException, IllegalStateException 등)

물론 도메인에서 중요한 부분은 예외를 재정의하는 것도 좋습니다.
이런 것들은 칼로 무 썰듯 그 선을 정하기가 어렵기 때문에 그때 그때 상황을 고려해서 구현하면 될 것 같습니다 :)

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * 원시 타입을 포장한 로또 번호 클래스
+ */
+public class LottoNumber {
```
값 객체 잘 구현해주셨네요 👍 
여유가 되신다면 final 키워드를 활용해서 불변 객체로 만들어보시는 것도 좋을 것 같습니다 :)
Comparable 인터페이스를 구현해봐도 재미있을 것 같네요.

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumbersMaker.java:
```diff
@@ -0,0 +1,33 @@
+		return IntStream.range(START_INCLUSIVE_NUMBER, END_EXCLUSIVE_NUMBER + 1)
+			.collect(ArrayList::new,
+				(pickedNumbers, number) -> pickedNumbers.add(new LottoNumber(number))
+				, (pickedNumbers1, pickedNumbers2) -> pickedNumbers1.addAll(pickedNumbers2));
```
지금도 괜찮긴한데요, 자주 쓰는 패턴은 아니라서 아래와 같이 변경해보시면 어떨까 싶습니다.

```java
.mapToObj(number -> new LottoNumber(number))
.collect(Collectors.toList();
```

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumbersMaker.java:
```diff
@@ -0,0 +1,33 @@
+		List<LottoNumber> lottoNumbers = new ArrayList<>(lottoNumberCandidates.subList(0,
+			SIZE_OF_LOTTERY_NUMBERS));  /*List의 subList는 deepCopy가 아니므로 새로 인스턴스를 생성하여 전달 */
+		Collections.sort(lottoNumbers,
+			(lottoNumberPre, lottoNumberPost) -> lottoNumberPre.getNumber() - lottoNumberPost.getNumber());
```
아래 방법으로 코드의 가독성과 의도를 개선해보면 어떨까요?
1. LottoNumber가 Comparable 인터페이스를 구현
2. LottoNumber에서 getNumber()한 결과로 Comparator.comparingInt

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -0,0 +1,43 @@
+		return lottoNumbers;
+	}
+
+	public int getNumOfNumbersInGroup() {
```
아마도 테스트 관점에서 작성된 코드가 아닐까 싶은데요,
LottoTicket이 가진 숫자의 개수를 가져와서 6개인지 확인하는 것보다는

생성자에서 불변식을 통해 6개의 숫자가 아니면 객체가 생성될 수 없도록 구현해보면 어떨까요?

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -0,0 +1,43 @@
+	private List<LottoNumber> lottoNumbers;
+
+	public LottoTicket(List<LottoNumber> lottoNumber) {
+		this.lottoNumbers = lottoNumber;
```
LottoNumber가 6개인지, 중복되지는 않는지에 대한 검사는 생성자에서 이루어지면 좋을 것 같습니다.

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	@Override
+	public String toString() {
+		StringBuilder lottoTicketToStrMaker = new StringBuilder();
+		lottoTicketToStrMaker.append('[');
+		lottoTicketToStrMaker.append(lottoNumbers.stream()
+			.map(lottoNumber -> lottoNumber.toString())
+			.collect(Collectors.joining(", ")));
+		lottoTicketToStrMaker.append(']');
+		return lottoTicketToStrMaker.toString();
```
잘 구현해주셨는데요 이렇게라도 가능할거에요~(지금 코딩을 못하는 상황이라 확인이..)
https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#joining-java.lang.CharSequence-java.lang.CharSequence-java.lang.CharSequence-

```java
lottoNumbers.stream()
	.map(lottoNumber -> lottoNumber.toString())
	.collect(Collectors.joining(", ", "[", "]"));
```

popotec replied:
오.. joining에 앞 뒤로도 붙일 수 있는 걸 확인을 안 해봤었네요. 감사합니다!

추가적으로 한가지 궁금한점이.. 
stream에서 String으로 형변환 하는 과정에서 String 객체 생성비용(lottoNumber.toString())이 조금 신경 쓰입니다.

이걸 줄이고자 collect 안에서 StringBuilder를 supplier로 주고,
lottoNumber를 char 형으로 변환하여 StringBuilder에 하나씩 append할까 생각도 해보았는데 구분자도 같이 들어가다 보니 코드가 좀 복잡해지더라구요..
이렇게 StringBuilder로 하는게 성능상 의미가 있을까요? 

지금은 데이터가 적기 때문에 String이나 StringBuilder가 큰 차이가 없지만, 
대용량 데이터를 이런식의 문자열로 만드는 경우라면 StringBuilder 하나로 만드는게 나을까요?

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicketIssuer.java:
```diff
@@ -0,0 +1,82 @@
+		return IntStream.range(0, numOfTickets)
+			.collect(ArrayList::new,
+				(tickets, i) -> tickets.add(new LottoTicket()),
+				(tickets1, tickets2) -> tickets1.addAll(tickets2));
```
이 부분도 위에서 언급했던 것처럼 더 쉽게 바꿔볼 수 있을 것 같아요.

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicketIssuer.java:
```diff
@@ -0,0 +1,82 @@
+		int numOfTickets = getNumOfTicketsToBuy(inputAmount);
+		List<LottoTicket> lottoTickets = issuTickets(numOfTickets);
+		return new LottoTickets(lottoTickets);
+	}
```
코드가 위에서 아래로 자연스럽게 읽히면 더 좋을 것 같습니다.
그런 의미에서 getNumOfTicketsToBuy 메서드도 buyTickets 아래 위치하는게 어떨까요?

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	public LottoTicket() {
+		this(LottoNumbersMaker.makeLottoNumbers());
```
LottoTicket은 단순히 값 객체라고 볼 수 있는데 값을 생성하는 로직이 담겨있는 LottoNumbersMaker에 강하게 의존하는 코드인 것 같아요.
빈 생성자는 private으로 닫고 인자를 받아서만 초기화할 수 있도록 하면 좋을 것 같습니다.

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatchResult.java:
```diff
@@ -0,0 +1,67 @@
+	private Map<LottoMatchRank, Integer> initResultSet() {
+		return Arrays.stream(LottoMatchRank.values()).collect(() -> new LinkedHashMap<LottoMatchRank, Integer>(),
+			(matches, rank) -> matches.put(rank, 0),
+			(matches1, matches2) -> matches1.putAll(matches2));
```
Collectors.toMap에 대해 알아보시면 더 짧게 코드를 구현해볼 수 있을거에요
(의도하신거라면 못본척해주세요..ㅎㅎ)

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatchResult.java:
```diff
@@ -0,0 +1,67 @@
+	/* 상금 합계 */
+	public long calculateProfit() {
+		return result.entrySet().stream()
+			.map(rank -> rank.getKey().getWinningMoney() * rank.getValue())
```
람다 인자가 rank인줄 알고 순간 헷갈렸네요.
entry가 더 적합한 이름이지 않을까 싶습니다. 아니면 요즘은 it도 많이 사용하는 것 같아요(이터레이터의 의미)

그리고 아래 mapToLong 대신 여기서 mapToLong 해도 되지 않을까요?

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatchResult.java:
```diff
@@ -0,0 +1,67 @@
+	private int getNumOfMatchResults() {
+		return result.entrySet()
+			.stream()
+			.map(matchResult -> matchResult.getValue())
```
여기도 한 번에 mapToInt가 가능해보여요

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatcher.java:
```diff
@@ -0,0 +1,44 @@
+
+	/* 당첨번호와 비교할 티켓정보 전달 */
+	public LottoMatchResult matchWinningAndTickets(LottoTickets lottoTickets) {
+		LottoMatchResult lottoMatchResult = new LottoMatchResult();
+		for (LottoTicket lottoTicket : lottoTickets.getLottoTickets()) {
+			setMatchResult(lottoMatchResult, lottoTicket);
+		}
+		return lottoMatchResult;
```
객체 외부에서 값을 set해주는 대신(add를 해주시긴 했지만요) 생성자에 lottoTickets와 lottoWinningNumbers를 넘기는 건 어떨까 싶습니다.
그렇게 되면 LottoMatcher와 LottoMatchResult의 역할이 중복되서 하나는 사라질 수도 있겠네요

minsungryu commented on src/main/java/lottoservice/matcher/LottoWinningNumbers.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	public static LottoWinningNumbers makeLottoWinningNumbers(List<Integer> numbers) {
+		validateSizeOfLotto(numbers);
```
이런 불변식은 정적 팩토리 메서드보다는 생성자 내부에 있는게 더 좋을 것 같습니다.

minsungryu commented on src/main/java/lottoservice/matcher/LottoWinningNumbers.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	private void validateHasNotDuplicateLottoNumber(Set<LottoNumber> lottoNumbers) {
+		if (lottoNumbers.size() < LottoNumbersMaker.SIZE_OF_LOTTERY_NUMBERS) {
```
이 조건문은 lottoNumbers가 7개 되는 오류는 못잡을 것 같아요.

minsungryu commented on src/main/java/lottoservice/matcher/LottoWinningNumbers.java:
```diff
@@ -0,0 +1,104 @@
+	public int compareWithNumbers(List<LottoNumber> ticketLottoNumbers) {
+		int matchCount = 0;
+		for (LottoNumber number : ticketLottoNumbers) {
+			matchCount += isMatchNumber(number) ? 1 : 0;
```
for-each 문 내부에서 부수효과를 일으키는 것보다 Stream API를 활용해보면 좋을 것 같습니다.

minsungryu commented on src/main/java/textcalculation/StringDelimiterParser.java:
```diff
@@ -0,0 +1,111 @@
+	private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile(CUSTOM_DELIMIERS_PATTERN);
+
+	protected static String[] splitText(String text) {
+		// Pattern 클래스의 인스턴스 생성 비용이 크기 때문에 매번 생성하는 대신 클래스 변수로 선언하고 재사용
```
👍

popotec replied:
메일로는 알람이 온 comment 부분이 github에는 안보여서.. 동일 메서드인 여기에 답글 남깁니다!

(관련 comment : ```protected로 선언하신 이유가 있을까요?```)

**StringDelimiterParser.splitText protected를 접근제어자로 사용한 이유** 
: StringDelimiterParser 클래스에서 사용하는 dlimiter 패턴들이 StringAddCalculator에 논리적으로 종속적이여서, 해당 패키지 내에서만 사용될것이니 public으로 하기 보다는 protected로 조금 더 제한해야겠다는 생각으로 사용하였습니다. 
=> 제가 맞게 생각한 걸까요~? 

그리고 관련해서 조금 더 생각해보니 CUSTOM_DELIMIERS_PATTERN와 BASIC_DELIMITERS를 상수 선언안하고 DI로 받아서 사용하게 되면 좀 더 유연하게 활용가능해서 public이 되어도 괜찮을것 같이 생각들었습니다.

그래서 이렇게 수정을 해보다가.. 이렇게 할 경우에는 Pattern을 인스턴스 멤버 변수로 관리하게 되면서 StringDelimiterParser를 생성할 때 마다 Pattern 객체 생성 비용을 치르게 되어 다시 돌아왔습니다^^;

더 좋은 방법이 있을까요?

그리고... 어제 늦은시간까지 리뷰해주신 것에 정말 감사합니다. 
덕분에 정말 많이 배우고 있습니다.

다른 comment 주신 것들도 고민해보며 보완해보겠습니다.

감사합니다!

minsungryu replied:
1. 네, 의도를 가지고 설계하셨다면 충분히 괜찮습니다 👍 
2. Pattern객체를 넘겨받는다기보다는 특정 전략을 수행하는 인터페이스에 의존하면 괜찮을 것 같아요. 어떤 구현체는 Pattern 객체를 내부적으로 활용할 수도 있는거고, 다른 구현체는 또 다른 방식으로 Parsing을 수행할 수도 있는거구요. 계산기 입장에서는 Parser가 어떻게 동작하는지보다는 결과값을 잘 넘겨주는 것에만 관심이 있을 것 같습니다.

minsungryu commented on src/main/java/textcalculation/StringDelimiterParser.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	private static boolean hasCustomDelimiter(Matcher m) {
+		if (m.find()) {
```
바로 `return m.find()` 해도 되겠는데요?

minsungryu commented on src/main/java/textcalculation/StringDelimiterParser.java:
```diff
@@ -0,0 +1,111 @@
+	private static final int SINGLE_LETTER_LENGTH = 1;
+	private static final Pattern CUSTOM_DELIMITER_PATTERN = Pattern.compile(CUSTOM_DELIMIERS_PATTERN);
+
+	protected static String[] splitText(String text) {
```
protected로 선언하신 이유가 있을까요?

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,55 @@
+			+ "사이 숫자만 가능합니다.";
+	private int number;
+
+	public LottoNumber(int number) {
```
값 객체는 그 수가 많아지면 메모리에 부담이 갈 수도 있을 것 같아요.
private으로 생성자를 막고 정적 팩토리 메서드를 통해 캐시된 객체를 반환하는 방법도 있습니다.
(IntegerCache 참고해보시면 좋을 것 같아요.)

이건 여유가 된다면 고려해보시는 것도 좋을 것 같습니다.

minsungryu commented on src/test/java/lottoservice/lottonumber/LottoNumbersMakerTest.java:
```diff
@@ -0,0 +1,26 @@
+	private static final int SIZE_OF_LOTTERY_NUMBERS = 6;
+
+	@Test
+	public void makeLottoNumbers_1에서45사이_6개의_숫자생성() {
```
제어할 수 없는 부분에 대해서 테스트하기 힘들지 않으신가요?
사실 위에서 많은 부분이 static으로 되어있어서 언급하려다가 적합한 테스트를 찾기 위해 보고 있었습니다.

랜덤, 시간 등 제어할 수 없는 부분에 대해서는 의존성 주입을 통해 해결하거나 인자로 받아서 해결할 수 있습니다.
따라서 NumberGenerator 또는 NumberSupplier와 같이 숫자를 제공하는 인터페이스를 통해 
프로덕션 코드에는 랜덤 생성을 하고 테스트 코드에서는 매뉴얼하게 생성해서 테스트를 진행할 수 있습니다.

LottoNumberMaker가 static 클래스가 아닌 외부 의존성을 주입받아서 로또 번호를 생성할 수 있도록 변경해보면 좋을 것 같습니다.

minsungryu commented on src/test/java/lottoservice/lottoticket/LottoTicketTest.java:
```diff
@@ -0,0 +1,29 @@
+	@Test
+	@RepeatedTest(10)
+	public void makeLottoTicket_로또_티켓_한장_발급() {
+		LottoTicket lottoTicket = new LottoTicket();
```
위에서 언급했듯이 LottoTicket이 값을 만드는 로직에 강하게 의존하고 있어서 발생하는 문제입니다.
테스트하기 어렵게 느껴지지 않으셨나요?

LottoTicket과 로직을 분리해보는걸 추천해드립니다.
사실 이런 테스트를 보면 프로덕션 코드 먼저 개발을 하고 테스트를 작성했다는 것도 알 수 있게 됩니다 🤣

minsungryu commented on src/test/java/lottoservice/matcher/LottoMatchResultTest.java:
```diff
@@ -0,0 +1,41 @@
+		LottoMatchResult lottoMatchResult = new LottoMatchResult();
+		lottoMatchResult.addMatchCount(LottoMatchRank.ONE_POINT);
+
+		Assertions.assertThat(lottoMatchResult.getResult().get(LottoMatchRank.SIX_POINT)).isEqualTo(0);
```
`get(...).get(...)` 보다는 `getResult(LottoMatchRank.SIX_POINT)`처럼 객체에 메시지를 보내는 것은 어떨까요?

minsungryu commented on src/test/java/lottoservice/matcher/LottoMatcherTest.java:
```diff
@@ -0,0 +1,57 @@
+		Assertions.assertThat(lottoMatchResult.calculateProfitPercentage()).isEqualTo(2.27);
+	}
+
+	private List<LottoNumber> getLottoNumbers(List<Integer> numbers) {
```
테스트 코드에서 로또 번호를 생성하기 위한 로직이 있는 것보다
LottoTicket 자체가 `List<Integer>`를 받아서 객체를 생성할 수 있도록 정적 팩토리 메서드를 제공하는 것은 어떨까요?

popotec replied:
조언해주신 부분을 참고하여 LottoTicket이 값 생성에 직접 관여하지 않고 값을 주입받도록 변경해보았습니다.
(설계를 하면서 무언가 찜찜한 부분들이 있었는데 디테일하게 달아주신 리뷰들 보고 많이 배우고 개선했습니다! 감사합니다!)

한가지 궁금한것은, 개선한 LottoTicket 코드에서 생성자 타입으로 List<LottoNumber>를 구현하였는데, List<Integer>를 인자로 받는 생성자나 정적 팩토리를 만들어도 될까요?
 
LottoTicket은 LottoNumber 리스트를 주입 받아서만 사용하고, Int 타입 List를 받아 LottoNumber 타입 리스트로 변환하는 일은 하지 않는 것이 도메인의 관심사를 분리해주지 않을까 생각했습니다. (List<LottoNumber> 생성 역할은 LottoNumbersMaker)

 (개선 전 코드 기준으로 조언해주신것 이어서, 지금 기준으로 여쭤보는 것이 또 다르긴 하지만 여쭤봅니다!)

minsungryu commented on src/test/java/lottoservice/matcher/LottoWinningNumbersTest.java:
```diff
@@ -0,0 +1,98 @@
+	@CsvSource({"3,34,22,17,26,7"})
+	public void makeLottoWinningNumbers_당첨번호_리스트_입력(ArgumentsAccessor argumentsAccessor) {
+		List<Integer> numbers = convertArgumentsToInteger(argumentsAccessor);
+		LottoWinningNumbers lottoWinningNumbers = LottoWinningNumbers.makeLottoWinningNumbers(numbers);
```
makeLottoWinningNumbers처럼 String을 받는 정적 팩토리 메서드를 구현해보는 것은 어떨까요?

minsungryu commented on src/test/java/lottoservice/matcher/LottoWinningNumbersTest.java:
```diff
@@ -0,0 +1,98 @@
+
+		List<LottoNumber> lottoNumbers = convertToLottoNumbers(numbers);
+		for (LottoNumber lottoNumber : lottoNumbers) {
+			assertThat(lottoWinningNumbers.getWinningNumbers().contains(lottoNumber)).isTrue();
```
`get(...).conatins(...)` 보다는 `hasNumber()`와 같이 객체에 메서드를 보내는 것이 어떨까요?

minsungryu reviewed (changes requested):
지희님 문자열 계산기도 수정하고 로또 미션도 구현하시느라 너무 고생이 많으셨습니다.
코멘트를 좀 남겼는데요 어렵게 느껴지는 부분이 있을 수 있습니다.
(남기고 보니 너무 많은..)
한 번 고민해보시고 다음 리뷰까지 모든 수정사항을 반영하지 않아도 되니 부담은 가지지 않으셨으면 좋겠습니다.
어려움이 있으면 언제든 편하게 DM 주세요

popotec committed:
refactor : 코드 간략화

불필요 if 체크 부분을 삭제하고 바로 리턴

popotec committed:
chore : 오타 수정 (pringNewLine -> printNewLine)

popotec committed:
refactor : 자바 제공 예외 클래스로 변경

기존 자바에서 제공하는 예외클래스로 충분히 의미있게 처리 가능하므로 불필요 예외 클래스를 삭제하고, IllegalStateException로 변경

popotec committed:
refactor : Comparable 인터페이스를 이용한 비

LottoNumbersMaker에서 생성한 로또 번호를 정렬할 때,
람다식으로 인터페이스를 구현하여 넘기지 않고, LottoNumber 클래스에서 Comparable 인터페이스를 구현하도록 변경
=> LottoNumber의 비교 논리를 LottoNumber에 둠으로서 보다 도메인 집약적

popotec committed:
chore(LottoNumber) : 래핑 대상 primitive 타입을 final로 선언

LottoNumber이 감싸고 있는 원시타입 number 값을 final로 선언하여 불변 성격을 부여.
=> 프로그램 복잡도가 커져 복잡한 로직을 거쳐도 값이 임의로 변경되지 않음을 보장

popotec committed:
chore(LottoNumbersMaker) : stream 간결하게 변경

collect 메서드 파라미터로 각 함수형 인터페이스를 구현하는 대신, stream에서 지원하는 다른 메서드를 이용해 간략하게 변경

popotec committed:
chore(LottoNumber) : 불필요한 외부 노출 메서드 접근제어자 변경

도메인 데이터의 값 검증 로직인 validateNumberInLottoRule를 외부 노출되지 않도록 private으로 변경

popotec committed:
chore(LottoTicketIssuer, LottoTicket) : stream 사용 간략화

popotec committed:
style(LottoTicketIssuer) : 코드 가독성을 위한 메서드 선후관계 변경

popotec committed:
refactor(LottoMatchResult) : Collectors.toMap을 이용해 스트림 간략화

popotec committed:
refactor(LottoMatchResult) : map, mapToInt 두개를 사용한 중개 연산을 mapToInt 하나만 사용하게 변경

popotec committed:
refactor(LottoWinningNumbers) : 객체의 인스턴스 멤버 변수에 대한 직접 참조 대신 객체에 메세지를 보내어 로또 번호 존재 확인

 LottoWinningNumbers의 인스턴스 멤버 변수 Set에 get으로 접근하여 contains 확인을 하기 보다는 hasMatchNumber 라는 메세지를 보내서 로또 번호가존재하는지 확인하도록 변경

popotec committed:
refactor(InputView, ResultView, ConsoleInputView, ConsoleResultView) : 입출력 구현 클래스의 action을 인터페이스로 추상화하여 분리

popotec committed:
refactor(LottoMain) : 추후 확장을 고려하여 입력, 출력 인스턴스를 DI.

popotec committed:
refactor(LottoWinningNumbers) : stream으로 로직 간결화

popotec committed:
refactor(LottoTicket) : LottoTicket에서 로또 번호 생성 로직 분리

LottoTicket은 단순히 LottoNumber 집합으로 이루어진 특성의 객체이므로, LottoTicket를 이루는 List<LottoNumber> 와의 결합을 제거하고 외부에서 주입받도록 함

popotec committed:
refactor(LottoNumbersMaker, LottoTicket, LottoWinningNumbers) : 역할 재분리

- LottoTicket과 LottoWinningNumbers은 일급컬렉션으로서의 성격을 강화하여 값 생성 로직을 분리하되, 해당 객체의 성격에 따른 유효성 검증 코드는 보완

- LottoNumbersMaker는 로또 번호를 리스트 값 생성의 역할을 강화함
=> 정적 팩토리 메서드를 이용해 세가지 방법으로 로또 번호 리스트 생성 기능을 제공 1)자동 생성, 2)Integer 리스트 형태로 입력 받아 생성, 3)구분자가 포함된 문자열 String 형식으로 받아 생성.

popotec committed:
refactor: 불필요 예외 클래스 삭제

popotec committed:
refactor(LottoWinningNumbers, LottoMatcher): LottoWinningNumbers가 티켓과의 비교 기능을 담당하도록 변경

LottoMatcher라는 별도의 유틸성 클래스를 만들기 보다는,
LottoWinningNumbers 객체에 로또 티켓 객체들을 전달하여 비교 결과를 요청하는 것으로 변경

popotec committed:
feat(LottoMatchResult): 랭크별 매칭 갯수 리턴

Map 객체에 직접 접근하여 get 하지 않고, LottoMatchResult에 메세지를 보내도록 변경

popotec committed:
test: LottoMatcherTest의 테스트를 LottoWinningNumbers로 옮김

popotec committed:
style: 포맷팅, 주석 추가

popotec committed:
test: LottoNumber comparable 테스트 추가

popotec committed:
test: LottoNumbersMaker 인자 전달 받아 객체 생성 테스트 추가

popotec committed:
feat(LottoTicket.hasLottoNumber): LottoTicket에 LottoNumber가 포함되어 있는지 체크

popotec committed:
feat: LottoTickets에 LottoTicket이 있는지 체크 구현

popotec committed:
refactor(LottoWinningNumbers): 객체에 메세지를 보내어 사이즈를 리턴하도록 변경

popotec committed:
chore: 메서드명 변경

popotec committed:
test: 테스트 가독성을 고려해 리팩토링

popotec committed:
chore: 미사용 상수 제거

popotec committed:
chore: 미사용 import 제거 및 포맷팅

popotec committed:
chore: 미사용 import 제거 및 포맷팅

popotec committed:
refactor: LottoNumber 캐싱 적용

LottoNumber 객체의 재생산을 방지하기 위하여 캐싱 적욕

popotec committed:
refactor: LottoNumber.valeOf로 적용

popotec committed:
chore: 미사용 import 삭제

popotec committed:
style: 포맷팅

minsungryu commented on src/main/java/lottoservice/LottoMain.java:
```diff
@@ -0,0 +1,89 @@
+	public static void main(String[] args) {
+		InputView userInputView = new ConsoleInputView();
+		ResultView resultView = new ConsoleResultView();
+		LottoMain lottoMain = new LottoMain(userInputView, resultView);
```
👍

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,84 @@
+	private final int number;
+	private static final LottoNumber[] cachedLottoNumbers;
+
+	static {
+		/* 스태틱 블록에서 LottoNumber 배열을 생성하여 number에 해당되는 index에 LottoNumber 객체를 캐싱 */
+		cachedLottoNumbers = new LottoNumber[LottoNumbersMaker.END_EXCLUSIVE_NUMBER + 1];    /*	편의를 위해 END_EXCLUSIVE_NUMBER +1개의 배열 생성 */
+		for (int i = 1; i <= LottoNumbersMaker.END_EXCLUSIVE_NUMBER; i++) {
+			cachedLottoNumbers[i] = new LottoNumber(i);
+		}
+	}
```
와 캐싱도 적용해보셨군요 멋지네요

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicketIssuer.java:
```diff
@@ -0,0 +1,83 @@
+
+	private static List<LottoTicket> issuTickets(int numOfTickets) {
+		return IntStream.range(0, numOfTickets)
+			.mapToObj((it) -> new LottoTicket(LottoNumbersMaker.makeLottoNumbers()))
```
`LottoNumbersMaker`의 제어 불가능한 부분(Collections.shuffle)에 의존하는 코드를 의존성 주입을 통해 해결해보면 좋을 것 같아요 :)

minsungryu reviewed (approved):
지희님, 로또 3단계 미션 수행하느라 고생많으셨습니다.
피드백도 잘 반영해주셨고 수많은 고민의 흔적이 느껴졌습니다.
바로 merge하겠습니다. 4단계도 화이팅입니다~
